src/server.c

```diff
@@ -1039,25 +1035,20 @@ void removeClientFromMemUsageBucket(client *c, int allow_eviction) {
  * all clients with roughly the same amount of memory. This way we group
  * together clients consuming about the same amount of memory and can quickly
  * free them in case we reach maxmemory-clients (client eviction).
- *
- * Note: This function filters clients of type no-evict, master or replica regardless
- * of whether the eviction is enabled or not, so the memory usage we get from these
- * types of clients via the INFO command may be out of date.
- *
- * returns 1 if client eviction for this client is allowed, 0 otherwise.
  */
-int updateClientMemUsageAndBucket(client *c) {
+void updateClientMemUsageAndBucket(client *c) {
     serverAssert(io_threads_op == IO_THREADS_OP_IDLE && c->conn);
+
+    /* Update client memory usage. */
+    updateClientMemoryUsage(c);
+
     int allow_eviction = clientEvictionAllowed(c);
     removeClientFromMemUsageBucket(c, allow_eviction);
 
     if (!allow_eviction) {
-        return 0;
+        return;
     }
 
-    /* Update client memory usage. */
-    updateClientMemoryUsage(c);
```

Comment:
I think this contradicts #11348.
@hpatro can you take a look?

Reply:
Yes, this reverts the change.

> The client memory usage and client-evict's mem_usage_bucket are not dependent on each other. Therefore, in the updateClientMemUsageAndBucket function, we can execute updateClientMemoryUsage for all clients (coming from the network) without checking clientEvictionAllowed.
This way, we can accurately track the memory usage of all clients in real-time, without clientsCron.

@soloestoy Active tracking of client's memory usage is a expensive operation if the client is already marked to be not evicted. I had introduced the change intentionally. Do you see any value of active tracking? It affects the TPS.

Reply:
There may be some differences in our understanding. From my perspective, real-time statistics of client memory and client eviction are not directly related. Client memory tracking belongs to the observability aspect, not only for client eviction. This is crucial for monitoring the running state of Redis.

Although the current `clientsCron` periodically collects statistics, periodic sampling can lead to significant deviations in the results. For example, if the memory statistics are not immediately updated after the client has sent all the data from the output buffer, it can result in the released memory still being accounted for, even exceeding the `used_memory`. Consequently, the `used_memory_dataset` may appear as negative (where dataset = used_memory - overhead, and overhead includes the memory of all clients). We have encountered such situations. Therefore, precise real-time statistics are essential. 

BTW, about the client eviction, I highly recommend enabling `maxmemory-clients`, as it is crucial for the stable operation of Redis and the host. It helps to avoid machine OOM situations caused by high traffic. Currently, in our platform, we configure it as 100% by default, and I also hope that it will be enabled by default in Redis 8.0.

Reply:
introspection is indeed different than eviction, but that's not the point.
the point is whether or not we want to compromise performance for introspection.
traditionally in redis, we compromised the accuracy of the metrics in favor of better performance, so i'm not sure if we wanna change that, certainly not without showing some benchmarks, and maybe try to optimize it.

regarding client eviction being enabled by default, i suppose it never occurred to me that setting it to 100% can be beneficial. i always thought of it as a low threshold designed protect the data from being evicted, and since no default number could be the right one, we kept it disabled by default. but here you make an argument that you use it to protect the host, not the data, so maybe that's a change we can make.
do you wanna make a PR to start a separate discussion on that?